---
title: "2 - RNA Analysis - Grouping and Standardization"
author: "David Alberto García Estrada"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Grouping and Normalization

Now we will proceed to a grouping and normalization from the complete counts table generated.
Let's read the table and store it in the counts variable.
It is actually a copy of the table from which "table" was generated.

```{r}
counts <- read.table("Counts_from_HISATSort-HTseq.txt", header = TRUE, 
                     row.names = 1, sep = "\t", comment.char = "")
```

We rename the headers as "Results_IPT27.txt".

```{r}
# We see the original name of each of the columns of the table of counts
names(counts)
names(counts) <- sub("Results_", "", names(counts)) 
names(counts) <- sub(".txt", "", names(counts)) 
# We reorder the columns as follows.
counts <- counts[,c("WT17", "WT27", "WT37",
                 "IPT17", "IPT27", "IPT37",
                 "WTCK17", "WTCK27", "WTCK37",
                 "WT110", "WT210", "WT310",
                 "IPT110", "IPT210", "IPT310",
                 "WTCK110", "WTCK210", "WTCK310")]
```

We filter out genes with a low number of reads as this can lead to errors in the subsequent statistical analysis.

```{r}
# First we see the number of genes we originally have. 
Original <- dim(counts) # Tells us that we have 12024 in 18 samples
# We filter by indicating that everything in each sample that has at least 2 reads is left
# at least 2 reads and that the sum of all the rows of that IDs has at least 5 reads
counts <- counts[rowSums(cpm(counts) >= 2) >=5,] 
Filtering <- dim(counts) # After filtering we are left with 5963 in 18 samples.
print(paste("The percentage remaining after filtering is",
            Filtering[1]/Original[1]*100)) 
# Percentage remaining is 59.5557034695066
```

Firts, we create a folder where the graphics are stored.

```{r}
Outpath <- "Dispersion/"
dir.create(Outpath, showWarnings = FALSE)
```

We check the relationship between pairs of library replicates to observe dispersion. 
This can be done with any pair of libraries.

```{r}
# The names of each of the libraries
names(counts)

# Comparison graphs between replicas
for (i in names(counts)){
  #i <- c("WT17") 
  for (j in names(counts)){
    #j <- c("IPT17")
    if (i != j){
      plot <- ggplot(data = counts, aes(x = log2(cpm(counts[,i])), 
                                        y = log2(cpm(counts[,j])))) +
        theme_classic() +
        labs(x=paste0("Counts ", i, "( Log2)"), 
             y=paste0("Counts", j, "(Log2)"),
            title = substitute(paste("Dispersion between replicates"))) +
        theme(legend.position = "right", 
              legend.text = element_text(size=12, face="bold"),
              axis.text = element_text(color = "black", size = 12, face = "bold"),
              axis.title = element_text(color = "black", size = 12, face = "bold"),
              plot.title = element_text(hjust = 0.5, size=15, face="bold")) +
        scale_color_brewer(palette = "Set1") +
        geom_point(size=3) +
        geom_smooth(formula = y ~ x, method="lm", 
                    color = "firebrick", fill = "firebrick") +
        #stat_fit_deviations(formula = y ~ x, colour = "black") +
        stat_poly_eq(aes(label = paste(stat(eq.label), 
                                       stat(adj.rr.label), sep = "*\", \"*")),
                     formula = y ~ x)
      print(plot) 
      svg(paste0(Outpath, "Dispersión_", i, "vs", j, ".svg"), width = 15, height = 15)
      print(plot) 
      dev.off()
      pdf(paste0(Outpath, "Dispersión_", i, "vs", j, ".pdf"), width = 15, height = 15)
      print(plot) 
      dev.off()
    }
  }
}

# To compare replicates in specific replace j and i by the replicate of interest.
i <- c("WT17") # substitute for replicate of interest
j <- c("IPT17") # substitute for replicate of interest
plot <- ggplot(data = counts, aes(x = log2(cpm(counts[,i])), 
                                  y = log2(cpm(counts[,j])))) +
  theme_classic() +
  labs(x = paste0("Counts ", i, " (Log2)"), 
       y = paste0("Counts", j, "(Log2)"), 
       title = substitute(paste("Dispersion between replicates"))) +
  theme(legend.position = "right", 
        legend.text = element_text(size=12, face="bold"),
        axis.text = element_text(color = "black", size = 12, face = "bold"),
        axis.title = element_text(color = "black", size = 12, face = "bold"),
        plot.title = element_text(hjust = 0.5, size=15, face="bold"))+
  scale_color_brewer(palette = "Set1") +
  geom_point(size=3) +
  geom_smooth(formula = y ~ x, method="lm", 
              color = "firebrick", fill = "firebrick") +
  #stat_fit_deviations(formula = y ~ x, colour = "black") +
  stat_poly_eq(aes(label =paste(stat(eq.label), 
                                stat(adj.rr.label), sep = "*\", \"*")),
              formula = y ~ x)
print(plot)
```

Continued on 3 - RNA Analysis - Differentially Expressed Genes